### DISCUSSION TITLE
Add custom error types to the standard library

### DISCUSSION THREAD

**Discussion by spaceface777** [vlang/v#11834] [Ideas]
Now that custom error types have been implemented into the language (see #8924 and #9043), the next step is to actually *use* these in the standard library.

For example:
 - the e.g. `return error('file `$file` not found')` calls inside the `os` module can be replaced with a new `FileNotFoundError` type
 - the generic `error_with_code` calls in `vlib/net/errors.v` should be replaced with a proper dedicated type for each error code
 - etc., etc.

It might be a good idea to implement these in batches (e.g. one PR could add proper errors to the `net` module, another could document part of `os`...)

**Comment by spaceface777** [comment]
Please note that every `error()` / `error_with_code()` call in vlib should not necessarily be replaced with a custom error type instance. The default error implementation exists for a reason, and is good enough for many cases. Custom errors are most useful in functions that can fail in several different ways, and figuring out what actually caused the error may be important to people.

**Comment by atakanyenel** [comment]
Hello, I had been following V for a while and this indeed looks like a "Good First issue". I want to take a look at this, and if I can, create the first batch of corrections. Feel free to assign this to me, and I will unassign myself, if I can't contribute to this by the end of this week.

**Comment by medvednikov** [comment]
Good luck and thanks @atakanyenel

**Comment by atakanyenel** [comment]
I came up with this example but I'm not sure. Can you give me an approval before I start changing the code base ?
```v
pub struct FileNotFound {
	msg  string = 'FileNotFound'
	code int
}

pub struct Unknown {
	msg  string = 'Unknown'
	code int
}

pub fn some_error(n int) ? {
	if n > 5 {
		return IError(FileNotFound{})
	} else {
		return IError(Unknown{})
	}
}

fn main() {
	some_error(10) or {
		match err {
			FileNotFound {
				println(err.msg)
			}
			else {
				println('else')
			}
		}
	}

	some_error(1) or {
		match err {
			Unknown {
				println(err.msg)
			}
			else {
				println('else')
			}
		}
	}
}
``` 
Creating the struct is fine but I'm not sure about the IError(FileNotFound{}) conversion.

**Reply by igotfr** [reply]
@atakanyenel what is the code field for?

**Reply by igotfr** [reply]
@atakanyenel why not specify the error types that can be returned?

```v
pub fn some_error(n int) ?(FileNotFound | Unknown) {
  if n > 5 {
    return IError(FileNotFound{})
  } else {
    return IError(Unknown{})
  }
}
```

**Reply by igotfr** [reply]
@atakanyenel or returning a value or error
```v
pub fn some_error(n int) int | ?(FileNotFound | Unknown) {
  if n > 0 {
    return 7
  }
  else if n > 5 {
    return IError(FileNotFound{})
  } else {
    return IError(Unknown{})
  }
}
```

**Reply by dumblob** [reply]
@atakanyenel and @cindRoberta mind that with RFC https://github.com/vlang/rfcs/pull/7 you'll have much easier to check for errors. I bet that designing IError interface(s) will be much more straightforward than now without the RFC.

Btw. do you know about any ways how to push the RFC from @danieldaeschle forward to be implemented in *V*? @danieldaeschle would you take the time and adjust the wording to accommodate for my last comment https://github.com/vlang/rfcs/pull/7#issuecomment-921706984 under the RFC?

**Reply by atakanyenel** [reply]
I'm not working on this anymore, so feel free to fix it if you think something is wrong.  

`code` field is a field defined in `IError` struct, so I needed it to cast it correctly. https://github.com/vlang/v/blob/47884dfd1ff75b5bc898866910710bb75931c7a6/vlib/builtin/option.v

V compiler didn't give any errors without the return types while I was working on this, (I don't know if it gives right now), so I didn't want to change function signature.

**Reply by danieldaeschle** [reply]
@dumblob I'm totally lost regarding that rfc. Could you make change requests in that PR? Then I will accept it.

**Comment by spaceface777** [comment]
I think you may need `IError(&FileNotFound{})` (due to a current compiler limitation), but it looks good to me otherwise.

Good luck! :)

**Comment by atakanyenel** [comment]
I went through all the modules and did the ones where an operation failed in multiple ways/was small enough component to not change big chunks of code. With this in my mind, only the remaining components remain:
- net
- os
- pg
- szip 
- v
- x
     - openssl
     - websocket

I will stop here for now(unassign myself) because there was a discussion about removing `err.code` from the error struct and I don't want program for a interface that might go away. Once that is finalized, I can implement that for errors.

**Comment by dumblob** [comment]
Just a tiny (but important) note - there are places, where errors are currently being ignored. But that should change. E.g. `os.File.close()` should return an error, because `C.fclose()` might fail with several different errors. See also https://www.joeshaw.org/dont-defer-close-on-writable-files/ .

**Comment by dumblob** [comment]
@medvednikov why was this moved to discussions? I think this is a clear actionable item. More so considering https://github.com/vlang/rfcs/pull/7 from @danieldaeschle will hopefully be implemented sooner or later.

**Comment by igotfr** [comment]
for reference:
https://github.com/vlang/v/discussions/11834
https://github.com/vlang/v/pull/9043
https://github.com/vlang/v/pull/8924
https://github.com/vlang/v/blob/47884dfd1ff75b5bc898866910710bb75931c7a6/vlib/builtin/option.v
